Follow up from fix trasient typo (#236)
socketry/async: guides/asynchronous-tasks/readme.md

```ruby
Async do
	3.times do |i|
		sleep 1 + i/10.0
		puts "Hello World #{i}"
	end
end
```

> This program prints "Hello World" 3 times. Before printing, it sleeps for 1, then 1.1, then 1.2 seconds. The total execution time is 3.3 seconds because the program executes sequentially.

Reviewer: Do you mind explaining the motivation to change this to `+ i/10.0`?

Author: Yes, I was trying to make the example more clear about when the stop happens. If all of the timers are firing at the 1 second mark, the timing of the stop is harder to follow.

But the tenths of a second are a bit cluttered for sure. I'd be happy to make those whole seconds or use some other approach for each thread to have a different timing value. Or I can put it back as it was if you'd prefer.

Reviewer: I see, I think maybe just `sleep(i)` can work and is a bit easier on the eye?

## Stopping a Task

> When a task completes execution, it will enter the `complete` state (or the `failed` state if it raises an unhandled exception).

Reviewer: All exceptions will cause the task to enter the failed state.

https://github.com/socketry/async/blob/main/lib/async/task.rb#L314-L317

Author: I thought you were saying that non-`StandardError`-derived exceptions would shut down the reactor and cause it to re-`raise`?

Reviewer: Yes, the task will still enter the failed state though in any case where an exception propagates up to the root of the fiber running the task. I suppose your change is okay because it's just mentioning one case, but both cases are true - i.e.:

> (or the `failed` state if it raises an unhandled `StandardError`-derived exception or any exception)

But I think we can just say "raises an unhandled exception" since in every case the task enters the `failed` state.

Reviewer: (If you think it's confusing, I'm also okay with your change, but it should be noted that any exception that propagates up causes the task to enter a `failed` state.)

end
```

> Unless your tasks all rescue and suppresses `StandardError`-derived exceptions, be sure to call ({ruby Async::Barrier#stop}) to stop the remaining tasks:

Reviewer: I think you can also suppress any exceptions and it's okay, even if it's questionable to do so.

## Exception Handling

> {ruby Async::Task} captures and logs exceptions. All unhandled `StandardError`-derived exceptions will cause the enclosing task to enter the `:failed` state. Non-`StandardError` exceptions are re-raised from the reactor and thus cause it to fail. This ensures that exceptions will always be visible and cause the program to fail appropriately.

Reviewer: ```suggestion
{ruby Async::Task} captures and logs exceptions. All unhandled exceptions will cause the enclosing task to enter the `:failed` state. Non-`StandardError` exceptions are re-raised immediately and will cause the reactor to exit. This ensures that exceptions will always be visible and cause the program to fail appropriately.
```

I suppose we could go with this. The exceptions are not reprised from the reactor - they are reprised from the fiber which generally causes the reactor to exit if it was the one that caused resume. Internally, it's more complex.

- A background worker or batch processing job which is independent of any specific operation, and is lazily created.
- A cache system which needs periodic expiration / revalidation of data/values.

Here is an example that keeps a cache of the current time string since that has only 1-second granularity
and you could be handling 1000s of requests per second.
The task doing the updating in the background is an implementation detail, so it is marked as `transient`.

```ruby
require 'async'
require 'thread/local' # thread-local gem.

class TimeStringCache
	extend Thread::Local # defines `instance` class method that lazy-creates a separate instance per thread
	
	def initialize
		@current_time_string = nil
	end
	
	def current_time_string
		refresh!
		
		return @current_time_string
	end
	
	private
	
	def refresh!
		@refresh ||= Async(transient: true) do
			loop do
				@current_time_string = Time.now.to_s
				sleep(1)
			end
		ensure

> 			# When the reactor terminates all tasks, this will be invoked.

Author: @ioquatix So this `ensure` block will be run because the reactor terminating the task will cause an exception to be raised from `sleep(1)`?

BTW do we prefer `sleep(1)` or `sleep 1` in this doc? I see both used here.

Reviewer: > So this ensure block will be run because the reactor terminating the task will cause an exception to be raised from sleep(1)?

Yes that's correct.

> do we prefer sleep(1) or sleep 1

I probably prefer `sleep(1)`.

Also, when comments refer to the next line, I prefer to end with a `:`.

Author: Ok, I've updated all uses of `sleep` to `sleep(...)` and terminated the comment (on the next line) with `:`.

				@current_time_string = Time.now.to_s
				sleep(1)
			end
		ensure
			# When the reactor terminates all tasks, this will be invoked:
			# When the reactor terminates all tasks, this will be invoked.
			# By clearing the variable, we ensure that the task will be recreated if needed again.
			@refresh = nil
		end
	end
end

Async do
	# If you are handling 1000s of requests per second, it can be an advantage to cache the current time string.
	p TimeCache.instance.current_time_string
	p TimeStringCache.instance.current_time_string
end
```

> Upon existing the top level async block, the {ruby @refresh} task will be set to `nil`. Bear in mind, you should not share these resources across threads; doing so would need some form of mutual exclusion.

Reviewer: I'm not sure if we should remove this statement, it should be clear that `@refresh` is bound to the current reactor per thread.

- A background worker or batch processing job which is independent of any specific operation, and is lazily created.
- A cache system which needs periodic expiration / revalidation of data/values.

> Bear in mind, in all of the above cases, you may need to validate that the background task hasn't been stopped, e.g. ???

Author: @ioquatix Can you suggest how to finish this sentence where the ??? is? I wasn't sure.

Reviewer: I think we can delete the sentence, IIRC, what I was trying to say, was that such tasks could be stopped due to going out of scope, rather than explicitly being `Task#stop`ed. In other words, `@refresh ||= Async(transient: true) {...}` is required to restart tasks if they go out of scope and have been stopped - "you may need to validate the background task hasn't been stopped".

A more specific example: let's say you had an in-process background job processing system. When you schedule a job, you should ensure it's running when adding the job. Otherwise, it may stop as a matter of program flow, and equally, it must be started as a matter of program flow.

This is captured in the below example.

Author: Sentence deleted.
